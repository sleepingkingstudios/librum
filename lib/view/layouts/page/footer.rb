# frozen_string_literal: true

module View::Layouts
  # Component for rendering the page footer.
  class Page::Footer < ViewComponent::Base; end
end
