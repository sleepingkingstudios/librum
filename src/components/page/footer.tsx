import * as React from 'react';

import { FancyHr } from '@components/fancy-hr';

export const PageFooter = (): JSX.Element => (
  <footer className="mt-3 shrink-0 text-center sm:text-left text-muted">
    <FancyHr className="mb-2" />

    What lies beyond the furthest reaches of the sky?
  </footer>
);
