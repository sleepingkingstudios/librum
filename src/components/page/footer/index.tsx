import * as React from 'react';

import { PageBreadcrumbs } from '../breadcrumbs';
import type { Breadcrumb } from '../breadcrumbs';

const defaultClassName = 'mt-3 shrink-0 text-center sm:text-left';

const renderBreadcrumbs = ({
  breadcrumbs = [],
}: {
  breadcrumbs?: Breadcrumb[] | JSX.Element,
}): JSX.Element => {
  if (!breadcrumbs) { return null; }

  if (!('length' in breadcrumbs)) { return breadcrumbs; }

  return (<PageBreadcrumbs breadcrumbs={breadcrumbs} />);
};

export const PageFooter = ({
  breadcrumbs = [],
}: {
  breadcrumbs?: Breadcrumb[] | JSX.Element,
}): JSX.Element => (
  <footer className={defaultClassName}>
    { renderBreadcrumbs({ breadcrumbs }) }

    <hr className="hr-fancy mb-2" />

    <span className="text-muted">
      What lies beyond the furthest reaches of the sky?
    </span>
  </footer>
);
