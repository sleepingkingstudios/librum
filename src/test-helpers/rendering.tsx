import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
