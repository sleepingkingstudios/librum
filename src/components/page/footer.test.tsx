import * as React from 'react';

import '@testing-library/jest-dom';

import { PageFooter } from './footer';
import { render } from '@test-helpers/rendering';

describe('<PageFooter>', () => {
  const theme = {
    textMuted: 'text-[#ff3366]',
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<PageFooter />, { theme });

    expect(asFragment()).toMatchSnapshot();
  });
});
