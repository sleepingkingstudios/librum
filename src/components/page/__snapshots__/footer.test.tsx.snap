// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageFooter> should match the snapshot 1`] = `
<DocumentFragment>
  <footer
    class="mt-3 shrink-0 text-center sm:text-left text-muted"
  >
    <hr
      class="fancy_hr border-text-strong after:border-text-strong dark:border-text-strong-dark dark:after:border-text-strong-dark mb-2"
    />
    What lies beyond the furthest reaches of the sky?
  </footer>
</DocumentFragment>
`;
